src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+            msg = f"monitoring task {uuid}, container {pname} under the rule ID {rid} targeting RSEs {rses} in rucio account {rucioAccount} is completely available"
+            logger.info(msg)
+            for rse in rses:
+                if rse not in doc['currentRSEs']:
+                    doc['currentRSEs'].append(rse)
+                    modify = True
+                    msg = f"update currentRSEs with {rse}"
+                    report.addEntry('monitoring', uuid, msg)
+        else:
+            # calculate the rule completion based on the 3 locks_* field
+            sumOfLocks = rdoc['locks_ok_cnt'] + rdoc['locks_replicating_cnt'] + rdoc['locks_stuck_cnt']
+            completion = rdoc['locks_ok_cnt'] / sumOfLocks
+            msg = f"monitoring task {uuid}, container {pname} under the rule ID {rid} targeting RSEs {rses} in rucio account {rucioAccount} has a fraction completion of {completion}"
+            logger.info(msg)
+            for rse in rses:
+                if rse in doc['currentRSEs']:
+                    doc['currentRSEs'].remove(rse)
+                    modify = True
+                    msg = f"delete rse {rse} from currentRSEs"
+                    report.addEntry('monitoring', uuid, msg)
+
+    # persist an up-to-date version of the pileup data structure in MongoDB
+    if modify:
+        logger.info(f"monitoring task {uuid}, update {pname}")
+        mgr.updatePileup(doc)
+        msg = f"update pileup {pname}"
+        report.addEntry('monitoring', uuid, msg)
+    else:
+        logger.info("monitoring task {uuid}, processed without update")
+
+
+def inactiveTask(doc, spec):
+    """
+    Perform single inactive task over provided MSPileup document
+
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    uuid = spec['uuid']
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    report = spec['report']
+    logger = spec.get('logger', getMSLogger(False))
+    pname = doc['pileupName']
+    if isinstance(rucioClient, Client):
+        inputArgs = {'scope': 'cms', 'name': pname, 'account': rucioAccount}
+        rules = [r for r in rucioClient.list_replication_rules(inputArgs)]
+    else:
+        rules = [r for r in rucioClient.listDataRules(pname)]
+    modify = False
+
+    for rdoc in rules:
+        # make a Rucio call to delete that rule id
+        rid = rdoc['id']
+        msg = f"inactive task {uuid}, container: {pname} for Rucio account {rucioAccount}, delete replication rule {rid}"
+        logger.info(msg)
+        if isinstance(rucioClient, Client):
+            rucioClient.delete_replication_rule(rid)
+            rses = [r['rse'] for r in rucioClient.list_rses(rdoc['rse_expression'])]
+        else:
+            rucioClient.deleteRule(rid)
+            rses = [rse for rse in rucioClient.evaluateRSEExpression(rdoc['rse_expression'])]
```

Comment:
Again, it already returns a list data type.

Reply:
done